aptos-move/block-executor/src/captured_reads.rs

```diff
@@ -0,0 +1,474 @@
+    fn contains(&self, other: &DataRead<V>) -> DataReadComparison {
+        let self_kind = self.get_kind();
+        let other_kind = other.get_kind();
+
+        if self_kind < other_kind {
+            DataReadComparison::Insufficient
+        } else {
+            let downcast_eq = if self_kind == other_kind {
+                // Optimization to avoid unnecessary clones (e.g. during validation).
+                self == other
+            } else {
+                self.as_kind(other_kind)
+                    .expect("Downcast to lower kind must succeed")
+                    == *other
+            };
+
+            if downcast_eq {
+                DataReadComparison::Contains
+            } else {
+                DataReadComparison::Inconsistent
+            }
+        }
+    }
+
+    /// If the reads contains sufficient information, extract this information and generate
+    /// a new DataRead of the desired kind (e.g. Metadata kind from Value).
+    pub(crate) fn as_kind(&self, kind: ReadKind) -> Option<DataRead<V>> {
+        let self_kind = self.get_kind();
+        if self_kind == kind {
+            return Some(self.clone());
+        }
+
+        (self_kind > kind).then(|| match (self, &kind) {
+            (DataRead::Versioned(_, v), ReadKind::Metadata) => {
+                DataRead::Metadata(v.as_state_value_metadata())
+            },
+            (DataRead::Versioned(_, v), ReadKind::Exists) => DataRead::Exists(!v.is_deletion()),
+            (DataRead::Resolved(_), ReadKind::Metadata) => DataRead::Metadata(Some(None)),
+            (DataRead::Resolved(_), ReadKind::Exists) => DataRead::Exists(true),
+            (DataRead::Metadata(maybe_metadata), ReadKind::Exists) => {
+                DataRead::Exists(maybe_metadata.is_some())
+            },
+            (_, _) => unreachable!("{:?}, {:?} must be covered", self_kind, kind),
+        })
+    }
+}
+
+/// Given a hashmap entry for a key, incorporate a new DataRead. This checks
+/// consistency and ensures that the most comprehensive read is recorded.
+fn update_entry<K, V: TransactionWrite>(
+    entry: Entry<K, DataRead<V>>,
+    read: DataRead<V>,
+) -> anyhow::Result<()> {
+    match entry {
+        Vacant(e) => {
+            e.insert(read);
+        },
+        Occupied(mut e) => {
+            let existing_read = e.get_mut();
+            match read.contains(existing_read) {
+                DataReadComparison::Contains => {
+                    *existing_read = read;
+                },
+                DataReadComparison::Inconsistent => {
+                    bail!("Inconsistent read results (due to speculation)")
+                },
+                DataReadComparison::Insufficient => {
+                    // New read is of a lower kind than existing read, but while
+                    // reading it must have been derived from the existing read.
+                    debug_assert!(
```

Comment:
oh, I didn't even consider INVARIANT_VIOLATION because we aren't inside the Move-VM so we can't directly throw those. We could instead log an error and do some fallback logic, but it may be too basic of an invariant to warrant that. Instead of the fallback, since we are in the read, we could return a fake STORAGE_ERROR to the transaction and cause it to abort, but it may need to be replaced to an INVARIANT_VIOLATION before it reaches the user. 

Not sure what makes the most sense, let me know your thoughts @igor-aptos?  tagging @zekun000 as well.

Reply:
While writing tests, I realized that if we assume proper caching, i.e. that the caller first checks if a read with higher kind exists to derive from (which we do, and makes sense to do), then neither "inconsistent" nor "insufficient" can be possible and can be asserted (not debug assert). However, if we don't make that assumption, we can bail!() in both cases, and the capture_read interface should forward the error to view to Move-VM. But we know we aren't supposed to be using it that way even during speculation, so we should also log an error.. But then returning Storage error we can't hope it will get cleared by speculation, so it'd have to be converted to invariant violation.

Honestly it seems the best course of action is in this case may be to log an error and fall back to sequential execution? We anyway were thinking of adding that functionality - I can easily add that e2e for this and then we can re-use for other cases. Should I?

Reply:
UPDATED: We aren't inside the Move-VM so we can't directly throw INVARIANT_VIOLATION. Since we are in the read, we could return a fake STORAGE_ERROR to the transaction and cause it to abort, but then the question is whether we need to replace the error somehow to more proper INVARIANT_VIOLATION before it reaches the user. @zekun000, @runtian-zhou - you would know best? Maybe Storage Error with appropriate message is sufficient?

If the caller first checks to derive from an existing read (which we do, and makes sense to do), then neither "inconsistent" nor "insufficient" can be possible. However, I think now that asserting is maybe not good, since if it isn't the intended way now or in the future (even because of the bug), surfacing it to trigger the assert would be highly unpredictable / dependent on luck/fate.

One other option is to log an error and fall back to sequential execution. This isn't hard to do and would be useful for other cases we had in mind. 
@igor-aptos @zekun000 what do you guys think?

Reply:
In general, I find reasonable to assert invariants that are localized and guaranteed to be true. (i.e. where looking at the local code we can be convinced it will never be triggered).

But here it is not really local - as it goes all the way to Unsync and Sync implementations, etc. 

For such - should be - but hard to guarantee, I would be throwing special CODE_INVARIANT_BROKEN issue, that can then trigger an alert, as well as fall back to sequential execution. I am trying to see how to do that right.

But here, even more importantly - is this even an invariant? Could it be perfectly possible that read of one kind returns to the call, then another transaction before it completes and writes output, and then read for a different kind is issued, which doesn't match up here? So bail is more appropriate (does bail here require transaction re-execution)?

 One stupid question, if transaction issues two reads, is there any caching, to make sure transaction sees the same thing, or will we end up in the Inconsistent case above?

Reply:
transaction session has its own cache, that one is fully local to it, but having multiple sessions have caused weird invariant behaviors in the past (but mainly speculative violations that re-execution was fixing).

the intended use (and should be the use after this change), is that every read will first come to CapturedReads and serve from here if possible (i.e. captured read exists with >= kind), if not do the full read logic and also capture it. It seems like there is one legit case when it can be inconsistent due to normal speculation - as you mentioned, when strictly lower kind is captured, and the higher kind is not consistent with it. 

I like your methodology a lot, let's see how to apply here. There seem to be two distinct behaviors: 
- given that inconsistency with strictly higher kind read can be speculative, in that case we can mark speculative failure (that will guarantee re-execution) and forward storage error (that will be cleared after re-execution).
- With inconsistency at the same kind of read, or having a read of a lower kind, let me alert/log an error and fallback to sequential execution. Sounds good?